File: back-end/app/models/device.py
"""
设备模型（已修复对齐前端字段）
修复：
1. type 字段枚举含 warning 相关设备类型
2. status 枚举含 'warning'
3. 字段别名：battery（前端）= battery_level（旧后端），signal = signal_strength
4. to_dict 输出前端期望字段：device_id, type, battery, signal, last_active
"""
from datetime import datetime
from app import db


class Device(db.Model):
    __tablename__ = 'devices'

    id               = db.Column(db.Integer, primary_key=True)
    device_id        = db.Column(db.String(50), unique=True, comment='设备编号如CAM-001')
    name             = db.Column(db.String(100), nullable=False)
    type             = db.Column(db.Enum(
                           'camera-visible','camera-infrared','fiber','smell',
                           'water-monitor','air-monitor','soil-monitor'
                       ), nullable=False)
    status           = db.Column(db.Enum('online','offline','warning','error'),
                                 default='offline')
    location         = db.Column(db.String(255), default='未知位置')
    latitude         = db.Column(db.Float)
    longitude        = db.Column(db.Float)
    battery          = db.Column(db.Integer, default=100, comment='电量0-100')
    signal_strength  = db.Column(db.Integer, default=100, comment='信号强度0-100')
    edge_node_id     = db.Column(db.String(50), comment='所属边缘节点ID')
    last_active      = db.Column(db.DateTime, comment='最后活跃时间')
    last_heartbeat   = db.Column(db.DateTime, comment='最后心跳时间')
    firmware_version = db.Column(db.String(50))
    hardware_version = db.Column(db.String(50))
    manufacturer     = db.Column(db.String(100))
    model            = db.Column(db.String(100))
    serial_number    = db.Column(db.String(100), unique=True)
    installation_date= db.Column(db.Date)
    maintenance_date = db.Column(db.Date)
    is_active        = db.Column(db.Boolean, default=True)
    created_at       = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at       = db.Column(db.DateTime, default=datetime.utcnow,
                                  onupdate=datetime.utcnow)

    def to_dict(self):
        """输出前端期望字段（device_id/type/battery/signal/last_active）"""
        return {
            'id':               self.id,
            'device_id':        self.device_id or str(self.id),
            'name':             self.name,
            'type':             self.type,
            'status':           self.status,
            'location':         self.location,
            'latitude':         self.latitude,
            'longitude':        self.longitude,
            'battery':          self.battery,           # 前端用 battery
            'battery_level':    self.battery,           # 兼容旧字段
            'signal':           self.signal_strength,   # 前端用 signal
            'signal_strength':  self.signal_strength,   # 兼容旧字段
            'edge_node_id':     self.edge_node_id,
            'last_active':      self.last_active.isoformat()    if self.last_active    else None,
            'last_heartbeat':   self.last_heartbeat.isoformat() if self.last_heartbeat else None,
            'firmware_version': self.firmware_version,
            'model':            self.model,
            'serial_number':    self.serial_number,
            'is_active':        self.is_active,
            'created_at':       self.created_at.isoformat() if self.created_at else None,
            'updated_at':       self.updated_at.isoformat() if self.updated_at else None,
        }

    def __repr__(self):
        return f'<Device {self.device_id or self.id}: {self.name}>'

File: back-end/src/models/Device.js
import { DataTypes } from 'sequelize';
import { sequelize } from '../config/database.js';

const Device = sequelize.define('Device', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  device_id: {
    type: DataTypes.STRING(50),
    allowNull: false,
    unique: true
  },
  name: {
    type: DataTypes.STRING(100),
    allowNull: false,
    validate: {
      len: {
        args: [1, 100],
        msg: '设备名称长度必须在1-100个字符之间'
      }
    }
  },
  type: {
    type: DataTypes.ENUM('camera-visible', 'camera-infrared', 'fiber', 'smell', 'water-monitor', 'air-monitor', 'soil-monitor'),
    allowNull: false
  },
  status: {
    type: DataTypes.ENUM('online', 'offline', 'warning', 'error'),
    defaultValue: 'offline'
  },
  location: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  latitude: {
    type: DataTypes.DECIMAL(10, 8),
    allowNull: false
  },
  longitude: {
    type: DataTypes.DECIMAL(11, 8),
    allowNull: false
  },
  battery: {
    type: DataTypes.INTEGER,
    validate: {
      min: 0,
      max: 100
    }
  },
  signal_strength: {
    type: DataTypes.INTEGER,
    validate: {
      min: 0,
      max: 100
    }
  },
  edge_node_id: {
    type: DataTypes.STRING(50)
  },
  last_active: {
    type: DataTypes.DATE
  },
  last_heartbeat: {
    type: DataTypes.DATE
  },
  firmware_version: {
    type: DataTypes.STRING(50)
  },
  hardware_version: {
    type: DataTypes.STRING(50)
  },
  manufacturer: {
    type: DataTypes.STRING(100)
  },
  model: {
    type: DataTypes.STRING(100)
  },
  installation_date: {
    type: DataTypes.DATEONLY
  },
  maintenance_date: {
    type: DataTypes.DATEONLY
  },
  is_active: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  }
}, {
  tableName: 'devices',
  timestamps: true,
  underscored: true,
  indexes: [
    { fields: ['device_id'], unique: true },
    { fields: ['type'] },
    { fields: ['status'] },
    { fields: ['latitude', 'longitude'] }
  ]
});

// 设备元数据子表
const DeviceMetadata = sequelize.define('DeviceMetadata', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  device_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'devices',
      key: 'id'
    }
  },
  ph: {
    type: DataTypes.DECIMAL(4, 2)
  },
  temperature: {
    type: DataTypes.DECIMAL(5, 2)
  },
  humidity: {
    type: DataTypes.DECIMAL(5, 2)
  },
  pm25: {
    type: DataTypes.DECIMAL(8, 2)
  },
  aqi: {
    type: DataTypes.INTEGER
  },
  heavy_metal: {
    type: DataTypes.DECIMAL(8, 4)
  },
  custom_data: {
    type: DataTypes.JSON
  },
  recorded_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  }
}, {
  tableName: 'device_metadata',
  timestamps: false,
  underscored: true,
  indexes: [
    { fields: ['device_id'] },
    { fields: ['recorded_at'] }
  ]
});

// 定义关联关系
Device.hasMany(DeviceMetadata, { foreignKey: 'device_id', as: 'metadata' });
DeviceMetadata.belongsTo(Device, { foreignKey: 'device_id' });

export { Device, DeviceMetadata };
export default Device;
